### ISSUE TITLE
[java] False positive on MissingStaticMethodInNonInstantiatableClass

### ISSUE THREAD

**Issue by jsotuyod** [pmd/pmd#590]
I've recently come across a bunch of false positives on this rule.

1. When looking for builder patterns it looks for public constructors, but fails when using the default one. The culprit is the line:

```
[count(./ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration/ConstructorDeclaration[@Public='true']) > 0]
```

which should be

```
[count(./ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration/ConstructorDeclaration[@Public='true']) > 0 or not(./ClassOrInterfaceBody/ClassOrInterfaceBodyDeclaration/ConstructorDeclaration)]
```

2. As part of the builder pattern detection (which shouldn't be needed and we can possibly get rid of), we check for a method returning an object defined on this very file, but we don't check if it's a nested type or not (shouldn't), and we require the use of the `this` qualifier, which is too weak to be of any use:

```
 and
  count(//ClassOrInterfaceDeclaration
            [@Nested='true']
            [@Static='true']
            [@Public='true']
            [.//MethodDeclaration
              [@Public='true']
              [.//ReturnStatement//AllocationExpression
                [ClassOrInterfaceType
                    [@Image = //ClassOrInterfaceDeclaration/@Image]
                ]
                [./Arguments//PrimaryPrefix/@ThisModifier='true']
              ]
            ]
       ) = 0
```

We should probably drop this last check altogether, improve the static builder detection, add a proper test and make sure we didn't break anything.

I'm away of the PC right now, so I'm creating this issue to keep track of it. I'll fix it as soon as I can.
